src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,609 @@
+}
+
+geometry_msgs::TwistWithCovarianceStamped
+        OpticalFlowEstimator::estimateVelocity(
+                const cv::Point2f& average_vec,
+                const tf2::Quaternion& curr_orientation,
+                const tf2::Quaternion& last_orientation,
+                const ros::Time& time) const
+{
+    // Get the pitch and roll of the camera in euler angles
+    // NOTE: CAMERA FRAME CONVENTIONS ARE DIFFERENT, SEE REP103
+    // http://www.ros.org/reps/rep-0103.html
+    double yaw, pitch, roll;
+    getYPR(curr_orientation, yaw, pitch, roll);
+
+    double last_yaw, last_pitch, last_roll;
+    getYPR(last_orientation, last_yaw, last_pitch, last_roll);
+
+    // Calculate time between last and current frame
+    double dt = (time - last_message_time_).toSec();
+
+    // Publish the orientation we're using for debugging purposes
+    iarc7_msgs::OrientationAnglesStamped ori_msg;
+    ori_msg.header.stamp = time;
+    ori_msg.data.pitch = pitch;
+    ori_msg.data.roll = roll;
+    ori_msg.data.yaw = yaw;
+    orientation_pub_.publish(ori_msg);
+
+    // Distance from the camera to the ground plane, along the camera's +z axis
+    double distance_to_plane = current_altitude_
+                             * std::sqrt(1.0
+                                       + std::pow(std::tan(pitch), 2.0)
+                                       + std::pow(std::tan(roll), 2.0));
+
+    // Converts measurements in pixels to measurements in meters in the plane
+    // parallel to the camera's xy plane and going through the point Pg, where
+    // Pg is intersection between the camera's +z axis and the ground plane
+    double current_meters_per_px = distance_to_plane
+                         / getFocalLength(target_size_,
+                                          flow_estimator_settings_.fov);
+
+    // Calculate the average velocity in the level camera frame (i.e. camera
+    // frame if our pitch and roll were zero)
+    // Note that cos(roll) is negative, because the z axis is down
+    cv::Point2f velocity_uncorrected;
+    velocity_uncorrected.x = (-average_vec.x * current_meters_per_px)
+                           / std::cos(pitch)
+                           / dt;
+    velocity_uncorrected.y = (-average_vec.y * current_meters_per_px)
+                           / -std::cos(roll)
+                           / dt;
+
+    double dp;
+    double dr;
+
+    if (last_pitch > CV_PI/2 && pitch < -CV_PI/2) {
+        dp = (pitch + 2*CV_PI - last_pitch);
+    } else if (last_pitch < -CV_PI/2 && pitch > CV_PI/2) {
+        dp = (pitch - last_pitch - 2*CV_PI);
+    } else {
+        dp = (pitch - last_pitch);
+    }
+
+    if (last_roll > CV_PI/2 && roll < -CV_PI/2) {
+        dr = (roll + 2*CV_PI - last_roll);
+    } else if (last_roll < -CV_PI/2 && roll > CV_PI/2) {
+        dr = (roll - last_roll - 2*CV_PI);
+    } else {
+        dr = (roll - last_roll);
+    }
+
+    double dpitch_dt = dp / dt;
```

Comment:
We don't do that because what we actually need is the average roll rate over the interval between frames, not the instantaneous roll rate, so this way is less noisy and more accurate

Reply:
Fair, but if our framerate were to approach the update rate of the kalmann filter this would be a problem.